internal/ls/lsutil/userpreferences.go

```diff
@@ -308,414 +249,505 @@ const (
+// For types which do not serialize as-is (tristate, enums, etc).
+var typeSerializers = map[reflect.Type]func(any) any{
+	reflect.TypeFor[core.Tristate](): func(val any) any {
+		switch val.(core.Tristate) {
+		case core.TSTrue:
+			return true
+		case core.TSFalse:
+			return false
+		default:
+			return nil
 		}
-	}
-	return QuotePreferenceUnknown
+	},
+	reflect.TypeFor[OrganizeImportsCollation](): func(val any) any {
+		if val.(OrganizeImportsCollation) == OrganizeImportsCollationUnicode {
+			return "unicode"
+		}
+		return "ordinal"
+	},
+	reflect.TypeFor[OrganizeImportsCaseFirst](): func(val any) any {
+		switch val.(OrganizeImportsCaseFirst) {
+		case OrganizeImportsCaseFirstLower:
+			return "lower"
+		case OrganizeImportsCaseFirstUpper:
+			return "upper"
+		default:
+			return "default"
+		}
+	},
+	reflect.TypeFor[OrganizeImportsTypeOrder](): func(val any) any {
+		switch val.(OrganizeImportsTypeOrder) {
+		case OrganizeImportsTypeOrderLast:
+			return "last"
+		case OrganizeImportsTypeOrderInline:
+			return "inline"
+		case OrganizeImportsTypeOrderFirst:
+			return "first"
+		default:
+			return "auto"
+		}
+	},
 }
 
-func (p *UserPreferences) Copy() *UserPreferences {
-	if p == nil {
-		return nil
-	}
-	prefCopy := *p
-	prefCopy.AutoImportSpecifierExcludeRegexes = slices.Clone(p.AutoImportSpecifierExcludeRegexes)
-	prefCopy.AutoImportFileExcludePatterns = slices.Clone(p.AutoImportFileExcludePatterns)
-	return &prefCopy
+type fieldInfo struct {
+	lowerName string   // lowercase Go field name for direct matching
+	aliases   []string // additional lowercase names (e.g., "provideprefixandsuffixtextforrename")
+	path      string   // dotted path for VS Code config (e.g., "preferences.quoteStyle")
+	fieldPath []int    // index path to field in struct
+	invert    bool     // whether to invert boolean values
 }
 
-func (p *UserPreferences) CopyOrDefault() *UserPreferences {
-	if p == nil {
-		return NewDefaultUserPreferences()
+var fieldInfoCache = sync.OnceValue(func() []fieldInfo {
+	var infos []fieldInfo
+	collectFieldInfos(reflect.TypeFor[UserPreferences](), nil, &infos)
+	return infos
+})
+
+// lowerNameIndex maps lowercase field names to fieldInfo index for O(1) lookup
+var lowerNameIndex = sync.OnceValue(func() map[string]int {
+	infos := fieldInfoCache()
+	index := make(map[string]int, len(infos)*2)
+	for i, info := range infos {
+		index[info.lowerName] = i
+		for _, alias := range info.aliases {
+			index[alias] = i
+		}
 	}
-	return p.Copy()
-}
+	return index
+})
 
-func (p *UserPreferences) ModuleSpecifierPreferences() modulespecifiers.UserPreferences {
-	return modulespecifiers.UserPreferences{
-		ImportModuleSpecifierPreference:   p.ImportModuleSpecifierPreference,
-		ImportModuleSpecifierEnding:       p.ImportModuleSpecifierEnding,
-		AutoImportSpecifierExcludeRegexes: p.AutoImportSpecifierExcludeRegexes,
-	}
-}
+func collectFieldInfos(t reflect.Type, indexPath []int, infos *[]fieldInfo) {
```

Comment:
I might have just made `collectFieldInfos` just return `[]fieldInfo` instead but this is probably fine.